cmd/kubeadm/app/phases/upgrade/compute.go

```diff
@@ -52,38 +52,64 @@ func (u *Upgrade) CanUpgradeKubelets() bool {
 type ClusterState struct {
 	// KubeVersion describes the version of the Kubernetes API Server, Controller Manager, Scheduler and Proxy.
 	KubeVersion string
 	// DNSVersion describes the version of the DNS add-on.
 	DNSVersion string
 	// KubeadmVersion describes the version of the kubeadm CLI
 	KubeadmVersion string
-	// KubeletVersions is a map with a version number linked to the amount of kubelets running that version in the cluster
-	KubeletVersions map[string]uint16
+	// The following maps describe the versions of the different components in the cluster.
+	// The key is the version string and the value is a list of nodes that have that version.
+	KubeAPIServerVersions         map[string][]string
+	KubeControllerManagerVersions map[string][]string
+	KubeSchedulerVersions         map[string][]string
+	EtcdVersions                  map[string][]string
+	KubeletVersions               map[string][]string
+}
+
+// AfterClusterState describes the state of certain versions for a cluster after an upgrade
+type AfterClusterState struct {
+	// KubeVersion describes the version of the Kubernetes API Server, Controller Manager, Scheduler, Proxy and Kubelet.
+	KubeVersion string
+	// DNSVersion describes the version of the DNS add-on.
+	DNSVersion string
+	// KubeadmVersion describes the version of the kubeadm CLI
+	KubeadmVersion string
 	// EtcdVersion represents the version of etcd used in the cluster
 	EtcdVersion string
 }
 
 // GetAvailableUpgrades fetches all versions from the specified VersionGetter and computes which
 // kinds of upgrades can be performed
-func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesAllowed, rcUpgradesAllowed, externalEtcd bool, client clientset.Interface, manifestsDir string, printer output.Printer) ([]Upgrade, error) {
+func GetAvailableUpgrades(versionGetterImpl VersionGetter, experimentalUpgradesAllowed, rcUpgradesAllowed bool, client clientset.Interface, printer output.Printer) ([]Upgrade, error) {
 	printer.Printf("[upgrade] Fetching available versions to upgrade to\n")
 
 	// Collect the upgrades kubeadm can do in this list
 	var upgrades []Upgrade
 
-	// Get the cluster version
-	clusterVersionStr, clusterVersion, err := versionGetterImpl.ClusterVersion()
+	// Get the kube-apiserver versions in the cluster
+	kubeAPIServerVersions, err := versionGetterImpl.ComponentVersions(kubeadmconstants.KubeAPIServer)
 	if err != nil {
 		return upgrades, err
 	}
+	if len(kubeAPIServerVersions) > 1 {
+		verMsg := []string{}
+		for version, nodes := range kubeAPIServerVersions {
+			verMsg = append(verMsg, fmt.Sprintf("%s on nodes %v", version, nodes))
+		}
+		klog.Warningf("Different API server versions in the cluster were discovered: %v. Please upgrade your control plane"+
+			" nodes to the same version of Kubernetes", strings.Join(verMsg, ", "))
+	}
+
+	// Get the lastest cluster version
+	clusterVersion, err := getLatestClusterVersion(kubeAPIServerVersions)
```

Comment:
just to check the logic.
we take the latest apiserver version as the cluster version, but we make sure to show a warning to the user if there are diff apiserver versions?

Reply:
L88-L95 has printed the message.  do you mean that moving the logic into the getLatestClusterVersion func?